File: android/app/src/main/kotlin/run/nuri/getagrip/ui/components/FingerPips.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.components

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.GripPosition
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.GetAGripTheme
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Motion
import run.nuri.getagrip.ui.theme.rememberReduceMotion

/// Four tappable bars — index to little — saying which fingers are on the edge, plus the
/// thumb's horizontal bar underneath.
///
/// It settles the app's worst ambiguity: "front 2" and "middle 2" are one word apart and two
/// grips, and a picture settles what a label flips a coin on. It composes any combination,
/// and under the no-emoji rule doubles as the app's iconography.
///
/// **ONE GLYPH, ONE RULE: full capsule at 22 : 38**, as `FingerGlyph` and `PalmHand` draw it —
/// the bar you tap while building must be the one hanging over the runner (Nuri, 2026-08-09).
@Composable
fun FingerPips(
    fingers: FingerSet,
    position: GripPosition,
    modifier: Modifier = Modifier,
    /// Drops the name line where the surface above already carries the grip name.
    showsName: Boolean = true,
    onChange: (FingerSet) -> Unit,
) {
    val palette = LocalGripPalette.current
    val haptics = LocalHapticFeedback.current

    // CLAMPED, the `ConsistencyCard` trap: uncapped, four bars plus the thumb want a whole
    // phone's width at the largest accessibility scale.
    val scale = LocalDensity.current.fontScale.coerceAtMost(1.4f)
    val width: Dp = (if (showsName) 34.dp else 28.dp) * scale
    val height: Dp = width * BAR_LENGTH_RATIO
    /// The tap target, not the drawing: a 34 dp bar is under the 44 dp floor.
    val hitWidth: Dp = maxOf(44.dp, width)

    /// A pinch IS thumb opposition, so the thumb is not a choice. `GripSpec` enforces it in the
    /// model, where a rule belongs: one living in a view is one the next view forgets.
    val locksThumb = position == GripPosition.pinch

    /// **The tick is TRIGGERED BY THE TAP, not by the value.** Keyed on `fingers`, choosing Pinch
    /// ticked twice: once in `PositionChipRow`, again when the model forced the thumb in. A tick
    /// has to name its cause.
    fun toggle(finger: FingerSet) {
        val next = FingerSelection.toggling(finger, fingers)
        if (next == fingers) return
        onChange(next)
        haptics.performHapticFeedback(HapticFeedbackType.SegmentTick)
    }

    Column(modifier, verticalArrangement = Arrangement.spacedBy(if (showsName) 8.dp else 6.dp)) {
        if (showsName) {
            // The canonical name, live, so picture and words are never two answers. A STATE change,
            // so `Motion.state` — see `Motion`.
    Crossfade(
        fingers.name,
        animationSpec = Motion.state(rememberReduceMotion()),
        label = "fingerName",
    ) { name ->
                Text(
                    name,
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = palette.inkPrimary,
                )
            }
        }

        Row {
            FingerSet.allFingers.forEachIndexed { index, finger ->
                val isOn = fingers.contains(finger)
                val shape = RoundedCornerShape(width / 2)
                Box(
                    Modifier
                        .size(hitWidth, height)
                        .clickable(
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                        ) { toggle(finger) }
                        .semantics {
                            role = Role.Button
                            contentDescription =
                                L10n.tr(
                            "%s finger, %s",
                            NAMES[index],
                            L10n.tr(if (isOn) "included" else "not included"),
                        )
                            selected = isOn
                        },
                    // BOTTOM-aligned: the tips, which are ON the edge, share a line.
                    contentAlignment = Alignment.BottomCenter,
                ) {
                    Box(
                        Modifier
                            .width(width)
                            .height(height * LENGTH_FACTOR[index])
                            .then(
                                if (isOn) {
                                    Modifier.background(palette.graphite, shape)
                                } else {
                                    // Stroked, not tinted: fill-vs-outline survives greyscale and Reduce Transparency.
                                    Modifier.border(1.dp, palette.inkTertiary.copy(alpha = 0.45f), shape)
                                },
                            ),
                    )
                }
            }
        }

        // The thumb as a horizontal bar under the fingers, where it sits in a pinch. Drawn vertical
        // it would be a short fifth finger, and the glyph must be read before any word.
        val thumbOn = fingers.hasThumb
        // The THICKEST digit, so thicker than a finger (a slim tab read as a mistake on the palm).
        val thumbThickness = width * 1.05f
        val thumbLength = hitWidth * 2 + 10.dp
        val thumbShape = RoundedCornerShape(thumbThickness / 2)
        Box(
            Modifier
                // The full pip-row width is the hit area.
                .size(hitWidth * 4 + 30.dp, maxOf(44.dp, thumbThickness))
                .then(
                    if (locksThumb) {
                        Modifier
                    } else {
                        Modifier.clickable(
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                        ) { toggle(FingerSet.thumb) }
                    },
                )
                .semantics {
                    role = Role.Button
                    contentDescription =
                        L10n.tr("Thumb, %s", L10n.tr(if (thumbOn) "included" else "not included"))
                    selected = thumbOn
                    if (locksThumb) disabled()
                },
            contentAlignment = Alignment.CenterStart,
        ) {
            Box(
                Modifier
                    .width(thumbLength)
                    .height(thumbThickness)
                    .then(
                        if (thumbOn) {
                            Modifier.background(
                                if (locksThumb) palette.graphite.copy(alpha = 0.55f) else palette.graphite,
                                thumbShape,
                            )
                        } else {
                            Modifier.border(1.dp, palette.inkTertiary.copy(alpha = 0.45f), thumbShape)
                        },
                    ),
            )
        }

        // **Said in WORDS**: the locked bar cannot say it by drawing. Disabling stops false press
        // feedback; this line stops "disabled" reading as "broken".
        if (locksThumb) {
            Text(
                tr("A pinch always includes the thumb."),
                style = MaterialTheme.typography.bodySmall,
                fontWeight = FontWeight.Medium,
                color = palette.inkTertiary,
            )
        }
    }
}

/// 22 × 38 dp, straight off the palm. The radius rule alone was not enough: at the old 40 × 52
/// a full capsule came out a fat oval. A drawing is its proportions as much as its corners.
private const val BAR_LENGTH_RATIO = 38f / 22f

/// Both hand pickers keep a real finger on the edge; the thumb cannot replace it.
internal object FingerSelection {
    fun toggling(finger: FingerSet, fingers: FingerSet): FingerSet {
        val next = if (fingers.contains(finger)) fingers.subtracting(finger) else fingers.union(finger)
        return if (next.subtracting(FingerSet.thumb).isEmpty) fingers else next
    }
}

/// A hand's proportions, not a bar chart's. Applied to the DRAWN bar only.
private val LENGTH_FACTOR = listOf(0.86f, 1.0f, 0.94f, 0.80f)

/// Index, middle, ring, little — `FingerSet.allFingers` order, so drawing and token agree.
/// A `get()`, not a stored list — see `Tab`.
private val NAMES: List<String>
    get() = listOf(L10n.tr("Index"), L10n.tr("Middle"), L10n.tr("Ring"), L10n.tr("Little"))

@Preview(name = "FingerPips", showBackground = true, widthDp = 360)
@Composable
private fun FingerPipsPreview() {
    GetAGripTheme {
        var fingers by remember { mutableStateOf(FingerSet.frontTwo) }
        Column(Modifier.padding(20.dp), verticalArrangement = Arrangement.spacedBy(24.dp)) {
            FingerPips(fingers, GripPosition.halfCrimp) { fingers = it }
            FingerPips(
                FingerSet.frontTwo.union(FingerSet.thumb),
                GripPosition.pinch,
            ) {}
        }
    }
}

File: Sources/UI/Routine/FingerPips.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// Four tappable bars — index to little — that say which fingers are on the edge.
///
/// "Front 2" and "middle 2" are one word apart and two grips; a picture settles what a
/// label flips a coin on. It beats a ten-chip grid (any combination, no enumeration) and
/// doubles as the app's iconography under the no-emoji rule.
///
/// Drawn as a hand, but each bar is TAPPED through a full-height container, so the target
/// stays ≥44pt where the bar is short.
struct FingerPips: View {
    @Binding var fingers: FingerSet
    var position: GripPosition
    /// Drops the name line where a navigation subtitle already carries the grip name and
    /// height is scarce.
    var showsName: Bool = true

    /// **The DRAWN bar, at the island hand's proportions** — see `barLengthRatio`. Fixed
    /// geometry, so it scales the sanctioned way rather than via a bare `.system(size:)`.
    ///
    /// **CLAMPED**, like `ConsistencyCard`'s dot: uncapped, four bars plus the thumb
    /// container (`hitWidth * 4 + 30`) want ~350 pt at accessibility3 against ~360 usable,
    /// on the max composer with no horizontal fallback. A picker that stops growing beats
    /// one that overflows offscreen.
    @ScaledMetric(relativeTo: .body) private var rawBarWidth: CGFloat = 34
    @ScaledMetric(relativeTo: .body) private var rawCompactBarWidth: CGFloat = 28
    private var barWidth: CGFloat { min(rawBarWidth, 48) }
    private var compactBarWidth: CGFloat { min(rawCompactBarWidth, 40) }

    /// 22 pt wide × 38 pt long, straight off `IslandHand`. The radius rule alone was not
    /// enough: at 40 × 52 a full capsule was a fat oval. A drawing is its proportions as
    /// much as its corners.
    private static let barLengthRatio: CGFloat = 38.0 / 22.0

    private var width: CGFloat { showsName ? barWidth : compactBarWidth }
    private var height: CGFloat { width * Self.barLengthRatio }
    /// The tap target, NOT the drawing: a 34 pt bar is under the 44 pt floor.
    private var hitWidth: CGFloat { max(44, width) }

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// Same order as `FingerSet.allFingers`, so the drawing never disagrees with the token.
    private static let names = [String(localized: "Index"), String(localized: "Middle"),
                                 String(localized: "Ring"), String(localized: "Little")]
    var body: some View {
        VStack(alignment: .leading, spacing: showsName ? 8 : 6) {
            // The canonical name, live, so the picture and the words are one answer.
            if showsName {
                Text(fingers.name)
                    .font(.system(.subheadline, weight: .semibold))
                    .foregroundStyle(Ink.primary)
                    .animation(Motion.state(reduceMotion),
                               value: fingers)
            }

            HStack(spacing: 10) {
                // By index: the bit and its spoken NAME are parallel arrays kept in step.
                ForEach(FingerSet.allFingers.indices, id: \.self) { index in
                    pip(index: index, finger: FingerSet.allFingers[index])
                }
            }
            .accessibilityElement(children: .contain)

            thumbBar

            // Said in words, because the locked bar cannot. Disabling stops the false
            // press feedback; this line stops "disabled" reading as "broken".
            if locksThumb {
                Text("A pinch always includes the thumb.")
                    .font(.system(.caption, weight: .medium))
                    .foregroundStyle(Ink.tertiary)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        // TRIGGERED BY THE TAP, not by the value: on `fingers`, choosing Pinch
        // ticked twice (once for `position` in `PositionChipRow`, again when the
        // model forces the thumb in). A tick names its cause, and one tap caused one.
        .sensoryFeedback(.selection, trigger: tapTick)
    }

    /// Bumped by `toggle` so the haptic fires for a FINGER TAP and nothing else.
    @State private var tapTick = 0

    /// A pinch IS thumb opposition, so under it the thumb is not a choice — see
    /// `GripSpec.fingers`, which enforces the same rule in the model.
    private var locksThumb: Bool { position == .pinch }

    private func pip(index: Int, finger: FingerSet) -> some View {
        let isOn = fingers.contains(finger)
        let name = Self.names[index]

        return Button {
            toggle(finger)
        } label: {
            ZStack(alignment: .bottom) {
                // The full-size container is the hit area; without it the little finger's
                // target would be 42pt.
                Color.clear
                RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                    .fill(isOn ? Accent.graphite : Color.clear)
                    .overlay {
                        if !isOn {
                            RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                                .strokeBorder(Ink.tertiary.opacity(0.45), lineWidth: 1)
                        }
                    }
                    // A hand's proportions (`HandGeometry`), on the DRAWN bar only.
                    .frame(width: width, height: height * HandGeometry.lengthFactor[index])
            }
            .frame(width: hitWidth, height: height)
            .contentShape(.rect)
        }
        .buttonStyle(PressFeedbackButtonStyle())
        .accessibilityLabel(String(localized: "\(name) finger, \(isOn ? String(localized: "included") : String(localized: "not included"))"))
        .accessibilityAddTraits(isOn ? [.isSelected] : [])
    }

    /// **A FULL CAPSULE, like `FingerGlyph` and the island hand.** The bar you tap while
    /// building must be the object you see hanging off the island while training (Nuri,
    /// 2026-08-09); a control with its own dialect of the app's only glyph feels assembled.
    private var cornerRadius: CGFloat { width / 2 }

    /// A set with no fingers on the edge is not a grip, so tapping the last engaged pip
    /// is a no-op rather than an empty state to recover from.
    /// The thumb, as a horizontal bar under the fingers, where a pinching thumb sits. Not a
    /// fifth column: drawn vertical, a thumb is just a short fifth finger.
    private var thumbBar: some View {
        let isOn = fingers.hasThumb
        // Spans the first two columns — the side a thumb actually opposes from.
        let thumbLength = hitWidth * 2 + 10
        // The THICKEST digit, so thicker than a finger — as in `IslandHand`, where a
        // slim tab beside four fat bars read as a mistake.
        let thumbThickness = width * 1.05

        return Button {
            toggle(.thumb)
        } label: {
            ZStack(alignment: .leading) {
                Color.clear
                // A `Capsule`: its half-height is the radius, whatever the length.
                Capsule()
                    .fill(isOn ? Accent.graphite : Color.clear)
                    .overlay {
                        if !isOn {
                            Capsule().strokeBorder(Ink.tertiary.opacity(0.45), lineWidth: 1)
                        }
                    }
                    .frame(width: thumbLength, height: thumbThickness)
            }
            // The full pip-row width, 44pt tall: the drawn bar alone is too small a target.
            .frame(width: hitWidth * 4 + 30, height: max(44, thumbThickness), alignment: .leading)
            .contentShape(.rect)
        }
        .buttonStyle(PressFeedbackButtonStyle())
        .disabled(locksThumb)
        .accessibilityLabel(String(localized: "Thumb, \(isOn ? String(localized: "included") : String(localized: "not included"))"))
        .accessibilityHint(locksThumb ? String(localized: "A pinch always includes the thumb") : "")
        .accessibilityAddTraits(isOn ? [.isSelected] : [])
    }

    private func toggle(_ finger: FingerSet) {
        let next = FingerSelection.toggling(finger, in: fingers)
        guard next != fingers else { return }
        fingers = next
        tapTick += 1
    }
}

/// Both hand pickers keep a real finger on the edge; the thumb cannot replace it.
enum FingerSelection {
    static func toggling(_ finger: FingerSet, in fingers: FingerSet) -> FingerSet {
        let next = fingers.contains(finger) ? fingers.subtracting(finger) : fingers.union(finger)
        return next.intersection(.four).isEmpty ? fingers : next
    }
}
